File: Java/ByTier/F/2103_F_Maximize_Nor.java
// Problem: CF 2103 F - Maximize Nor
// https://codeforces.com/contest/2103/problem/F

/**
 * Problem: Maximize Nor
 * Description:
 *   For each index i in an array of k-bit integers, find the maximum bitwise nor 
 *   among all subarrays containing index i. Bitwise nor is computed cumulatively 
 *   from left to right.
 *
 * Algorithms/Techniques:
 *   - Segment Tree with lazy propagation for range updates and queries.
 *   - Efficient sliding window approach using a list of (value, position) pairs.
 *   - For each new element, update the list of possible values and apply range 
 *     updates in the segment tree to track maximum nor values.
 *
 * Time Complexity: O(n * k * log n), where n is array length and k is number of bits
 * Space Complexity: O(n)
 */
import java.io.*;
import java.util.*;
public class Main {
    static class Node {
        int max, tag;
    }
    static class RangeTree {
        int size;
        Node[] tree;
        RangeTree(int n) {
            size = n;
            tree = new Node[4 * n];
            for (int i = 0; i < tree.length; i++) tree[i] = new Node();
        }
        void apply(int id, int v) {
            // Update the maximum value at node if v is greater
            if (v > tree[id].max) tree[id].max = v;
            // Propagate update to children
            if (v > tree[id].tag) tree[id].tag = v;
        }
        void push(int id) {
            int v = tree[id].tag;
            // If tag is not zero, propagate it to children
            if (v != 0) {
                apply(id << 1, v);
                apply(id << 1 | 1, v);
                tree[id].tag = 0;
            }
        }
        void cover(int id, int l, int r, int ql, int qr, int v) {
            // No overlap
            if (ql > r || qr < l) return;
            // Complete overlap - apply value and stop recursion
            if (ql <= l && r <= qr) {
                apply(id, v);
                return;
            }
            push(id);
            int m = (l + r) >> 1;
            cover(id << 1, l, m, ql, qr, v);
            cover(id << 1 | 1, m + 1, r, ql, qr, v);
            // Update current node's max value
            tree[id].max = Math.max(tree[id << 1].max, tree[id << 1 | 1].max);
        }
        void cover(int l, int r, int v) {
            cover(1, 0, size - 1, l, r, v);
        }
        void collect(int id, int l, int r, int[] res) {
            // Leaf node reached
            if (l == r) {
                res[l] = tree[id].max;
                return;
            }
            push(id);
            int m = (l + r) >> 1;
            collect(id << 1, l, m, res);
            collect(id << 1 | 1, m + 1, r, res);
        }
    }
    static class Pair {
        int val, pos;
        Pair(int v, int p) { val = v; pos = p; }
    }
    public static void main(String[] args) throws Exception {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(in.readLine());
        StringBuilder out = new StringBuilder();
        while (T-- > 0) {
            StringTokenizer st = new StringTokenizer(in.readLine());
            int n = Integer.parseInt(st.nextToken()), k = Integer.parseInt(st.nextToken());
            int mask = (1 << k) - 1; // Create a mask for k bits
            int[] a = new int[n];
            st = new StringTokenizer(in.readLine());
            for (int i = 0; i < n; i++) a[i] = Integer.parseInt(st.nextToken());
            RangeTree rt = new RangeTree(n);
            List<Pair> prev = new ArrayList<>();
            for (int r = 0; r < n; r++) {
                List<Pair> curr = new ArrayList<>();
                // Add current element
                curr.add(new Pair(a[r], r));
                for (Pair p : prev) {
                    // Compute nor of pair (p.val, a[r])
                    int x = ~(p.val | a[r]) & mask;
                    boolean seen = false;
                    for (Pair q : curr) {
                        if (q.val == x) {
                            // Update position of existing value to earlier index
                            if (p.pos < q.pos) q.pos = p.pos;
                            seen = true;
                            break;
                        }
                    }
                    if (!seen) curr.add(new Pair(x, p.pos));
                }
                for (Pair q : curr) {
                    // Update range in segment tree with maximum value from subarray
                    rt.cover(q.pos, r, q.val);
                }
                prev = curr;
            }
            int[] ans = new int[n];
            rt.collect(1, 0, n - 1, ans);
            for (int v : ans) out.append(v).append(' ');
            out.append('\n');
        }
        System.out.print(out);
    }
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByTier/F/2103_F_Maximize_Nor.py
# Problem: CF 2103 F - Maximize Nor
# https://codeforces.com/contest/2103/problem/F

# ==================================================================
# Problem: F. Maximize Nor
# Purpose: For each index i in an array, find the maximum bitwise nor
#          among all subarrays containing that index.
#
# Algorithms:
# 1. Preprocessing using cumulative NOR operations.
# 2. Bit manipulation techniques to efficiently compute results.
# 3. Dynamic programming for optimization with sliding windows.
#
# Time Complexity: O(n * k) per test case, where n is the array length and k is number of bits.
# Space Complexity: O(n) for storing arrays and auxiliary data structures.
# ==================================================================


import sys

input = sys.stdin.readline


def fg():
    return int(input())


def fgh():
    return [int(xx) for xx in input().split()]


def tt():
    sys.stdout.flush()


def sd():
    print("Yes")


def df():
    print("No")


step = [1]
for i in range(20):
    step.append(step[-1] * 2)


def nor(x, b):
    global k
    return (2**k - 1) ^ (x | b)  # Compute bitwise NOR of x and b


def nor1(a):
    if len(a) == 1:
        return a[0]
    return nor(nor1(a[: len(a) - 1]), a[-1])  # Recursively compute cumulative NOR


for __ in range(fg()):
    n, k = fgh()
    a = fgh()
    if n == 1:
        print(a[0])
        continue
    ans = a[:]  # Initialize result array with input values

    # First pass: Update ans[i] based on neighboring elements
    for i in range(n - 1):
        ans[i] = max(ans[i], nor(a[i], a[i + 1]))
        ans[i + 1] = max(ans[i + 1], nor(a[i + 1], a[i]))

    # Precompute prefix NORs and update maximum from left to right
    pr = [a[0]]
    for i in range(1, n):
        pr.append(nor(pr[-1], a[i]))
    mx = pr[-1]
    for i in range(n - 1, -1, -1):
        mx = max(mx, pr[i])
        ans[i] = max(ans[i], mx)

    # Precompute suffix NORs and update maximum from right to left
    pr = [a[1]]
    for i in range(2, n):
        pr.append(nor(pr[-1], a[i]))
    mx = pr[-1]
    for i in range(n - 1, 0, -1):
        mx = max(mx, pr[i - 1])
        ans[i] = max(ans[i], mx)

    # Iterate over all bits for optimization using specific bit groups
    for i in range(n - 1):
        for j in range(k):
            # If current element has bit j set, process accordingly
            if a[i] & step[j]:
                # Compute prefix NORs starting from index i
                pr = [a[i]]
                L = -1
                for l in range(i + 1, n + 1):
                    if l == n or (a[l] & step[j]):
                        L = l
                        break
                    pr.append(nor(pr[-1], a[l]))
                mx = pr[-1]
                # Propagate maximums backward through selected ranges
                for o in range(L - 1, i - 1, -1):
                    mx = max(mx, pr[o - i])
                    ans[o] = max(ans[o], mx)
                # Special case handling when L > i+1
                if L > i + 1:
                    pr = [a[i + 1]]
                    for l in range(i + 2, n + 1):
                        if l == n or (a[l] & step[j]):
                            break
                        pr.append(nor(pr[-1], a[l]))
                    mx = pr[-1]
                    for o in range(L - 1, i, -1):
                        mx = max(mx, pr[o - i - 1])
                        ans[o] = max(ans[o], mx)
                # Further refinement when L > i+2
                if L > i + 2:
                    pr = [a[i + 2]]
                    for l in range(i + 3, n + 1):
                        if l == n or (a[l] & step[j]):
                            break
                        pr.append(nor(pr[-1], a[l]))
                    mx = pr[-1]
                    for o in range(L - 1, i + 1, -1):
                        mx = max(mx, pr[o - i - 2])
                        ans[o] = max(ans[o], mx)
    print(*ans)


# https://github.com/VaHiX/codeForces/